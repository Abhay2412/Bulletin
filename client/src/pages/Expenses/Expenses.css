a {
    text-decoration-color:  rgba(104, 57, 13, 0.54);
}


.expensesTitle {
    text-align: left;
    color:#68390D;
    padding-top: 10px;

}

.expensesSubtitles {
    text-align: left;
    color:#68390D;
    padding-top: 20px;
}

.expensesTable {
    display: flex;
    flex-direction: column;
    padding-left: 40px;
    padding-right: 40px;
    padding-top: 20px;
}

.expensesTableRow {
    display: flex;
    flex-direction: row;
    justify-content: space-evenly;
    margin-bottom: 10px;
    align-items: center;
    height: 35px;    
    color: #68390D;
}

.tableCellsFormatting{
    color: #68390D;
    background-color: #F6F0E9;
    border-right: 1px solid rgba(104, 57, 13, 0.24);
    border-left: 1px solid rgba(104, 57, 13, 0.24);
    box-shadow: 2px 4px rgb(196, 192, 192, 0.5);
}

.addNewBox {
    text-align: left;
    width: 90%;
}

.calculateTotal {
    width: 90%;
    justify-content: right;
    padding-right: 20px;
}

.overlay {
    position: absolute;
    display: flex;
    height: 100%;
    width: 100%;
    align-items: center;
    justify-content: center;
    margin: 0 !important;
    z-index: 1; 
    
}

.overlayBox {
    width: 80%;
    height: 60%;
    background-color: white;
    text-align: left;
    padding-left: 5%;
    padding-top: 2%;
    color: #B8A590;
    border: 1px solid black;
}

.overlayHeader {
    display: flex;
    justify-content: space-between;
    padding-right: 2%;
}

.overlayFields {
    color: #675443;
    padding-left: 5%;
    padding-top: 2%;
}

.overlayFieldRow {
    display: flex;
    justify-content: space-between;
    padding-right: 50%;
    align-items: center;
}

.saveDiv {
    display: flex;
    justify-content: right;
    padding-top: 5%;
    padding-right: 10%;
}

.saveButton {
    background-color: #AA896B !important;
    color: white !important;
    width: 30%;
}
