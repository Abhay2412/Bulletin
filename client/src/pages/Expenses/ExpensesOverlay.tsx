import { TextField, Typography, Select, Button } from '@mui/material'
import React from 'react'
import "./ExpensesOverlay.css"

interface Props {
    header: string,
    type: string,
    nameField: string,
    dateType: string,
    balanceField: string,
}

export default function ExpensesOverlay({type, header, nameField, dateType, balanceField}: Props) {
  return (
    <div className="overlay">
        <div className="overlayBox">
            <Typography variant="h3">
                {header}
            </Typography>

            <div className="overlayFields">
                <div className="overlayFieldRow">
                    <Typography variant="h6">
                        {type} Name
                    </Typography>
                    <TextField variant='standard' label={nameField} />
                </div>

                <div className="overlayFieldRow">
                    <Typography variant="h6">
                        {dateType} Date
                    </Typography>
                    <TextField variant='standard' label={nameField} />
                </div>
                <div className="overlayFieldRow">
                    <Typography variant="h6">
                        Assigned To
                    </Typography>
                    <Select>
                        
                    </Select>
                </div>
                <div className="overlayFieldRow">
                    <Typography variant="h6">
                        Balance
                    </Typography>
                    <TextField variant="standard" label={balanceField}/>
                </div>
                <div className="saveDiv">
                    <Button className="saveButton">
                        Save {type}
                    </Button>
                </div>
            </div>

        </div>
    </div>
  )
}
