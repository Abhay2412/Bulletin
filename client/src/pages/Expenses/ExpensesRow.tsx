import React from "react";
import { useState } from "react";
import { Box, Button, Select, MenuItem } from "@mui/material";

interface Props {
  name: string;
  date: string;
  assigned: string;
  balance: number;
}

const mockMemberInfo = [
  { id: 1, name: "Liane Doe", email: "liane.doe@gmail.com", role: "Admin" },
  { id: 2, name: "Dad Doe", email: "dad.doe@gmail.com", role: "Admin" },
  { id: 3, name: "Logan Doe", email: "logan.doe@gmail.com", role: "Member" },
  { id: 4, name: "Aly Doe", email: "aly.doe@gmail.com", role: "Member" },
];

export default function ExpensesRow({ name, date, assigned, balance }: Props) {

  return (
    <>
      <Box className="expensesTableRow tableCellsFormatting" width="20%">
        {name}
      </Box>
      <Box className="expensesTableRow tableCellsFormatting" width="15%">
        {date}
      </Box>

      <Box className="expensesTableRow tableCellsFormatting" width="35%">
        {assigned}
      </Box>
      <Box className="expensesTableRow tableCellsFormatting" width="20%">
        ${balance.toFixed(2)}
      </Box>
      <Box className="expensesTableRow" width="5%">
        <Button>X</Button>
      </Box>
      <Box className="expensesTableRow" width="5%">
        <Button sx={{ textDecoration: "underline" }}>edit</Button>
      </Box>
    </>
  );
}
