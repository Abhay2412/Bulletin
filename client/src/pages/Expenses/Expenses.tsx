import React, { useEffect, useState } from "react";
import { Typography, Box, Button } from "@mui/material";
import "./Expenses.css";
import ExpensesRow from "./ExpensesRow";
import ExpensesOverlay from "./ExpensesOverlay";
import { useParams } from "react-router-dom";
import axiosInstance from "../../axios";
import Budget from "../../models/Budget";
import Expense from "../../models/Expense";

export default function Expenses() {
  const params = useParams();
  const [boardName, setBoardName] = useState();
  const [expensesTotal, setExpensesTotal] = useState(0);
  const [budgetTotal, setBudgetTotal] = useState(0);
  const [popupState, setPopupState] = useState(false);
  const [popupHeader, setPopupHeader] = useState("New Expense");
  const [popupType, setPopupType] = useState("Expense");
  const [popupDateType, setPopupDateType] = useState("Due");
  const [expenses, setExpenses] = useState<Expense[]>([]);
  const [budgets, setBudgets] = useState<Budget[]>([]);

  useEffect(() => {
    setExpensesTotal(0);

    expenses.map((expenses) => {
      setExpensesTotal((expensesTotal) => expensesTotal + expenses.amount);
    });
  }, [expenses]);

  useEffect(() => {
    setBudgetTotal(0);

    budgets.map((budget) => {
      setBudgetTotal((budgetTotal) => budgetTotal + budget.balance);
    });
  }, [budgets]);

  const newExpensePopup = () => {
    setPopupState(true);
    setPopupHeader("New Expense");
    setPopupType("Expense");
    setPopupDateType("Due");
  };

  const newBudgetPopup = () => {
    setPopupState(true);
    setPopupHeader("New Budget");
    setPopupType("Budget");
    setPopupDateType("End");
  };

  useEffect(() => {
    if (!params.board_id) {
      return;
    }

    axiosInstance
      .get("/getExpenses", { params: { id: params.board_id } })
      .then((res) => {
        console.log("expenses repsonse is: ", res);
        setExpenses([...res.data.expenses]);
      })
      .catch((err) => {
        console.log("error getting user expenses: ", err);
      });

    axiosInstance
      .get("/getBudgets", { params: { id: params.board_id } })
      .then((res) => {
        console.log("budgets repsonse is: ", res);
        setBudgets([...res.data.budgets]);
      })
      .catch((err) => {
        console.log("error getting user budgets: ", err);
      });
  }, [params.board_id]);

  useEffect(() => {
    axiosInstance
      .get("/getBoard", { params: { id: params.board_id } })
      .then((res) => {
        setBoardName(res.data.board.data.name);
        console.log("Information recieved Successfully");
      })
      .catch((err) => {
        console.log("error getting user boards: ", err);
      });
  }, [params.board_id]);

  return (
    <>
      {popupState ? (
        <ExpensesOverlay
          type={popupType}
          header={popupHeader}
          setExpenses={setExpenses}
          dateType={popupDateType}
          setPopupState={setPopupState}
          boardId={params.board_id}
          isEdit={false}
          setBudgets={setBudgets}
        />
      ) : null}
      <div style={{ marginLeft: "5%" }}>
        <a href={"/board/"+params.board_id}>
          <Typography
            variant="subtitle1"
            color="primary.light"
            align="left"
            sx={{ pt: "1vh" }}
          >
            Back to {boardName} Board - Main
          </Typography>
        </a>

        <Typography className="expensesTitle" variant="h3">
          Expenses
        </Typography>

        <Box>
          <Typography
            className="expensesSubtitles"
            fontWeight={"bold"}
            variant="h6"
          >
            Bills & Expenses
          </Typography>

          <Box className="expensesTable">
            <Box className="expensesTableRow">
              <Box width={"20%"}>
                <Typography fontWeight={"bold"}>Expenses Name</Typography>
              </Box>
              <Box width={"15%"}>
                <Typography fontWeight={"bold"}>Due Date</Typography>
              </Box>
              <Box width={"35%"}>
                <Typography fontWeight={"bold"}>Assigned To</Typography>
              </Box>
              <Box width={"20%"}>
                <Typography fontWeight={"bold"}>Balance</Typography>
              </Box>
              <Box width={"5%"}></Box>
              <Box width={"5%"}></Box>
            </Box>
            {expenses.map((expense) => {
              return (
                <Box className="expensesTableRow">
                  <ExpensesRow
                    name={expense.name}
                    date={expense.deadline}
                    assigned={expense.assignee}
                    balance={expense.amount}
                    type={"Expense"}
                    boardId={params.board_id}
                    expenseId={expense.id}
                    setExpenses={setExpenses}
                    expenses={expenses}
                    budgets={budgets}
                    setBudgets={setBudgets}
                  ></ExpensesRow>
                </Box>
              );
            })}
            <Box className="expensesTableRow tableCellsFormatting addNewBox">
              <Button
                variant="text"
                fullWidth={true}
                onClick={() => newExpensePopup()}
              >
                <Typography>+ Add New Expense</Typography>
              </Button>
            </Box>
            <Box className="expensesTableRow calculateTotal">
              <Typography variant="h5">
                Total: ${expensesTotal.toFixed(2)}
              </Typography>
            </Box>
          </Box>
        </Box>

        <Box>
          <Typography
            className="expensesSubtitles"
            fontWeight={"bold"}
            variant="h6"
          >
            Budgets
          </Typography>

          <Box className="expensesTable">
            <Box className="expensesTableRow">
              <Box width={"20%"}>
                <Typography fontWeight={"bold"}>Budget Name</Typography>
              </Box>
              <Box width={"15%"}>
                <Typography fontWeight={"bold"}>End Date</Typography>
              </Box>
              <Box width={"35%"}>
                <Typography fontWeight={"bold"}>Assigned To</Typography>
              </Box>
              <Box width={"20%"}>
                <Typography fontWeight={"bold"}>Balance</Typography>
              </Box>
              <Box width={"5%"}></Box>
              <Box width={"5%"}></Box>
            </Box>
            {budgets.map((budget) => {
              return (
                <Box className="expensesTableRow">
                  <ExpensesRow
                    name={budget.name}
                    date={budget.date}
                    assigned={budget.assigned}
                    balance={budget.balance}
                    type={"Budget"}
                    boardId={params.board_id}
                    expenseId={budget.id}
                    setExpenses={setExpenses}
                    expenses={expenses}
                    budgets={budgets}
                    setBudgets={setBudgets}
                  ></ExpensesRow>
                </Box>
              );
            })}
            <Box className="expensesTableRow tableCellsFormatting addNewBox">
              <Button
                variant="text"
                fullWidth={true}
                onClick={() => newBudgetPopup()}
              >
                <Typography>+ Add New Budget</Typography>
              </Button>
            </Box>
            <Box className="expensesTableRow calculateTotal">
              <Typography variant="h5">
                Total: ${budgetTotal.toFixed(2)}
              </Typography>
            </Box>
          </Box>
        </Box>
      </div>
    </>
  );
}
