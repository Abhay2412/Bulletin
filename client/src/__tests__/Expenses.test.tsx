import React from "react";
import Expenses from "../pages/Expenses/Expenses";
import { configure, mount, ReactWrapper } from "enzyme";
import renderer from "react-test-renderer";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { store } from "../store";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";

configure({ adapter: new Adapter() });

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    board_id: "g7HGuvfnGe3nhQgAF8qY",
  }),
}));

describe("Testing <Expenses /> Component", () => {
  it("should render properly", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <Expenses />
          </BrowserRouter>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should open a modal", () => {
    const wrapper: ReactWrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <Expenses />
        </BrowserRouter>
      </Provider>
    );

    wrapper.find("button").find("#add-expense-button").simulate("click");
    expect(wrapper.find(".overlayBox"));
  });

  it("should open a modal", () => {
    const wrapper: ReactWrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <Expenses />
        </BrowserRouter>
      </Provider>
    );

    wrapper.find("button").find("#add-budget-button").simulate("click");
    expect(wrapper.find(".overlayBox"));
  });
});
