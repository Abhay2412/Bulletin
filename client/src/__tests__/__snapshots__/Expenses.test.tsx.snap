// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Expenses /> Component should render properly 1`] = `
<div
  style={
    Object {
      "marginLeft": "5%",
    }
  }
>
  <a
    href="/board/g7HGuvfnGe3nhQgAF8qY"
  >
    <h6
      className="MuiTypography-root MuiTypography-subtitle1 MuiTypography-alignLeft css-10x4mua-MuiTypography-root"
    >
      Back to 
       Board - Main
    </h6>
  </a>
  <h3
    className="MuiTypography-root MuiTypography-h3 expensesTitle css-gepadz-MuiTypography-root"
  >
    Expenses
  </h3>
  <div
    className="MuiBox-root css-0"
  >
    <h6
      className="MuiTypography-root MuiTypography-h6 expensesSubtitles css-tjvwag-MuiTypography-root"
    >
      Bills & Expenses
    </h6>
    <div
      className="expensesTable MuiBox-root css-0"
    >
      <div
        className="expensesTableRow MuiBox-root css-0"
      >
        <div
          className="MuiBox-root css-ufscxh"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Expenses Name
          </p>
        </div>
        <div
          className="MuiBox-root css-w56oqh"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Due Date
          </p>
        </div>
        <div
          className="MuiBox-root css-2rhtyh"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Assigned To
          </p>
        </div>
        <div
          className="MuiBox-root css-ufscxh"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Balance
          </p>
        </div>
        <div
          className="MuiBox-root css-ndbrsi"
        />
        <div
          className="MuiBox-root css-ndbrsi"
        />
      </div>
      <div
        className="expensesTableRow tableCellsFormatting addNewBox MuiBox-root css-0"
      >
        <button
          className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-10vjzfb-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          id="add-expense-button"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            + Add New Expense
          </p>
        </button>
      </div>
      <div
        className="expensesTableRow calculateTotal MuiBox-root css-0"
      >
        <h5
          className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          Total: $
          0.00
        </h5>
      </div>
    </div>
  </div>
  <div
    className="MuiBox-root css-0"
  >
    <h6
      className="MuiTypography-root MuiTypography-h6 expensesSubtitles css-tjvwag-MuiTypography-root"
    >
      Budgets
    </h6>
    <div
      className="expensesTable MuiBox-root css-0"
    >
      <div
        className="expensesTableRow MuiBox-root css-0"
      >
        <div
          className="MuiBox-root css-ufscxh"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Budget Name
          </p>
        </div>
        <div
          className="MuiBox-root css-w56oqh"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            End Date
          </p>
        </div>
        <div
          className="MuiBox-root css-2rhtyh"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Assigned To
          </p>
        </div>
        <div
          className="MuiBox-root css-ufscxh"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Balance
          </p>
        </div>
        <div
          className="MuiBox-root css-ndbrsi"
        />
        <div
          className="MuiBox-root css-ndbrsi"
        />
      </div>
      <div
        className="expensesTableRow tableCellsFormatting addNewBox MuiBox-root css-0"
      >
        <button
          className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-10vjzfb-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          id="add-budget-button"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            + Add New Budget
          </p>
        </button>
      </div>
      <div
        className="expensesTableRow calculateTotal MuiBox-root css-0"
      >
        <h5
          className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          Total: $
          0.00
        </h5>
      </div>
    </div>
  </div>
</div>
`;
